geoportailv3/static/js/query/querydirective.js

```diff
@@ -0,0 +1,242 @@
+      var buffer = 5;
+      var ll = ol.proj.transform(
+          map.getCoordinateFromPixel(
+          [evt.pixel[0] - buffer, evt.pixel[1] - buffer]),
+          map.getView().getProjection(), 'EPSG:2169');
+      var ur = ol.proj.transform(
+          map.getCoordinateFromPixel(
+          [evt.pixel[0] + buffer, evt.pixel[1] + buffer]),
+          map.getView().getProjection(), 'EPSG:2169');
+      var box = ll.concat(ur);
+
+      $http.get(
+          getInfoServiceUrl,
+          {params: {
+            'layers': layersList.join(),
+            'box': box.join()
+          }}).then(
+          goog.bind(function(resp) {
+            goog.array.forEach(resp.data, function(item) {
+              item['layerLabel'] = layerLabel[item.layer];
+            });
+            this['content'] = resp.data;
+            this['infoOpen'] = true;
+            this.clearFeatures_();
+            this.lastHighlightedFeatures_ = [];
+            for (var i = 0; i < resp.data.length; i++) {
+              this.lastHighlightedFeatures_.push.apply(
+                  this.lastHighlightedFeatures_,
+                  resp.data[i].features
+              );
+            }
+            this.highlightFeatures_(this.lastHighlightedFeatures_);
+          },this));
+    }
+  },this));
+
+
+  map.on('pointermove', goog.bind(function(evt) {
+    if (evt.dragging) {
+      return;
+    }
+    var pixel = map.getEventPixel(evt.originalEvent);
+    var hit = map.forEachLayerAtPixel(pixel, function(layer) {
+      if (goog.isDefAndNotNull(layer)) {
+        var metadata = layer.get('metadata');
+        if (goog.isDefAndNotNull(metadata)) {
+          if (goog.isDefAndNotNull(metadata['is_queryable']) &&
+              metadata['is_queryable']) {
+            return true;
+          }
+        }
+      }
+      return false;
+    });
+    map.getTargetElement().style.cursor = hit ? 'pointer' : '';
+  },this));
+};
+
+
+/**
+ * @private
+ */
+app.QueryController.prototype.clearFeatures_ = function() {
+  this.overlay_.getFeatures().clear();
+};
+
+
+/**
+ * provides the template path according with the fact
+ * that the template for the current layer is remote or not
+ * @param {{remote_template: boolean, template: string, layer: string}} layer
+ * @return {string} the template path.
+ * @export
+ */
+app.QueryController.prototype.getTemplatePath = function(layer) {
+  if (layer['remote_template'] === true) {
+    return (this.remoteTemplateUrl_ + '?layer=' + layer['layer']);
+  }
+  return (this.templatePath_ + '/' + layer['template']);
+};
+
+
+/**
+ * @param {Array<string>} features the features to highlight
+ * @private
+ */
+app.QueryController.prototype.highlightFeatures_ = function(features) {
+
+  var encOpt = {
+    dataProjection: 'EPSG:2169',
+    featureProjection: this['map'].getView().getProjection()
+  };
+
+  var jsonFeatures = (new ol.format.GeoJSON()).readFeatures({
+    'type': 'FeatureCollection',
+    'features': features},
+  encOpt
+  );
```

Comment:
> I don't find the doc about readFeaturesFromObject . Do you know where I can find it ?

It's [not](https://github.com/openlayers/ol3/blob/master/src/ol/format/geojsonformat.js#L425-L428) marked as `@api` in the OpenLayers 3 code, so it is not in the doc.

But we're different from other people :) Because we use the Closure Compiler, and compile the application and OpenLayers together. This means we have access to any OpenLayers symbol/function, including symbols that don't have the `@api` marker.

One could argue that using non-`@api` OpenLayers symbols is bad, because it means we use things that may disappear or change in the future. It's true, but the compiler helps us here. If something changes in OpenLayers the compiler will catch it immediately if we pay attention to correctly typing our code.